iroh/src/magicsock.rs

```diff
@@ -1561,7 +1633,50 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
-type RelayRecvResult = Result<(PublicKey, quinn_udp::RecvMeta, Bytes), io::Error>;
+/// Channel for [`MagicSock::poll_recv_relay`] to receive QUIC datagrams from relays.
+///
+/// This channel contains both the send and receive sides.  The receive side implements the
+/// poll interface with required wakers.
+#[derive(Debug)]
+struct RelayRecvChannel {
+    sender: mpsc::Sender<RelayReadResult>,
+    receiver: parking_lot::Mutex<mpsc::Receiver<RelayReadResult>>,
+    waker: parking_lot::Mutex<Option<Waker>>,
```

Comment:
we should just use `AtomicWaker` for this

Reply:
Postponed for followup.  Left in comment.